service-mvt/src/main/java/org/oskari/service/mvt/ToMVTSpace.java

```diff
@@ -0,0 +1,101 @@
+package org.oskari.service.mvt;
+
+import java.util.Arrays;
+
+import com.vividsolutions.jts.geom.Coordinate;
+import com.vividsolutions.jts.geom.Geometry;
+import com.vividsolutions.jts.geom.GeometryFactory;
+import com.vividsolutions.jts.geom.LineString;
+import com.vividsolutions.jts.geom.LinearRing;
+import com.vividsolutions.jts.geom.MultiPoint;
+import com.vividsolutions.jts.geom.Point;
+import com.vividsolutions.jts.geom.util.GeometryEditor.GeometryEditorOperation;
+
+/**
+ * Transform Geometry to Vector tile space
+ * Also snaps to integer grid and removes duplicate points
+ */
+public class ToMVTSpace implements GeometryEditorOperation {
+
+    private final double tx;
+    private final double ty;
+    private final double sx;
+    private final double sy;
+
+    public ToMVTSpace(double tx, double ty, double sx, double sy) {
+        this.tx = tx;
+        this.ty = ty;
+        this.sx = sx;
+        this.sy = sy;
+    }
+
+    public Geometry edit(Geometry geometry, GeometryFactory factory) {
```

Comment:
This returns a copy of the given geometry that is "cut" to fit in the tile "bbox"? And translates "proper" coordinates to pixel coordinates inside the tile?

Reply:
This translates the real-world coordinates of a geometry to the inner virtual-coordinate space of the tile. Inside the tile a coordinate consists of two integers (x,y) both in the range of [0,extent] (usually extent=4096).
Intersecting the geometry with the (real-world) spatial extent of the tile is a separate process, not connected to this operation.